File: mobile/www/js/domain/taste.js
// Анализ вкуса и рекомендации (тот же алгоритм, что в ПК-версии) — только вычисления.
import { norm, title } from "./titles.js";

const W = { completed: 2, watching: 1.5, planned: 0.6, postponed: 0.3, dropped: -2 };

/** Профиль вкуса по данным пользователя ({library, progress, anime} из хранилища). */
export function profile(s) {
  const weights = {};
  const counts = {};
  for (const [id, e] of Object.entries(s.library)) {
    let w = W[e.status] || 0;
    if (e.status) counts[e.status] = (counts[e.status] || 0) + 1;
    if (e.favorite) w += 3;
    if (e.score) w += (e.score - 5) / 2.5;
    weights[id] = (weights[id] || 0) + w;
  }
  for (const [id, eps] of Object.entries(s.progress)) {
    const watched = Object.values(eps).filter((v) => v.watched).length;
    weights[id] = (weights[id] || 0) + Math.min(watched, 12) / 8;
  }
  const g = {}, t = {};
  let ySum = 0, yW = 0;
  const titles = [];
  for (const [id, w] of Object.entries(weights)) {
    const a = s.anime[id];
    if (!a) continue;
    titles.push([a, w]);
    const gs = a.genres || [];
    for (const x of gs) g[x] = (g[x] || 0) + w / Math.sqrt(gs.length);
    if (a.type) t[a.type] = (t[a.type] || 0) + w;
    if (a.year && w > 0) { ySum += a.year * w; yW += w; }
  }
  const pos = Object.entries(g).filter(([, v]) => v > 0);
  const top = Math.max(1, ...pos.map(([, v]) => v));
  const genres = Object.fromEntries(pos.sort((a, b) => b[1] - a[1]).map(([k, v]) => [k, v / top]));
  const tpos = Object.entries(t).filter(([, v]) => v > 0);
  const tsum = tpos.reduce((a, [, v]) => a + v, 0) || 1;
  const types = Object.fromEntries(tpos.sort((a, b) => b[1] - a[1]).map(([k, v]) => [k, v / tsum]));
  const yearMean = yW ? ySum / yW : null;
  const liked = titles.filter(([, w]) => w >= 1.5).sort((a, b) => b[1] - a[1]).slice(0, 12).map(([a]) => a);
  const disliked = new Set(Object.entries(g).filter(([, v]) => v < -1).map(([k]) => k));
  return { genres, types, yearMean, liked, disliked, counts, empty: !titles.some(([, w]) => w > 0), titles };
}

/** Запросы каталога: любимые жанры по одному и парой + популярное и свежее. */
export function candidateQueries(p, byName) {
  const top = Object.keys(p.genres).filter((x) => byName[x]).slice(0, 4);
  const queries = top.map((x) => ({ genres: [byName[x]], sorting: "RATING_DESC", limit: 40 }));
  if (top.length >= 2) queries.push({ genres: [byName[top[0]], byName[top[1]]], sorting: "RATING_DESC", limit: 30 });
  queries.push({ sorting: "RATING_DESC", limit: 40 }, { sorting: "FRESH_AT_DESC", limit: 30 });
  return queries;
}

/** [{score, rel, reason}] по убыванию. */
export function score(p, rels) {
  const likedSets = p.liked.map((a) => [a, new Set(a.genres)]);
  const out = [];
  for (const rel of rels) {
    const gs = (rel.genres || []).map((x) => x.name);
    if (!gs.length) continue;
    const bad = gs.filter((x) => p.disliked.has(x)).length;
    if (bad && bad >= gs.length / 2) continue;
    let gScore = gs.reduce((a, x) => a + (p.genres[x] || 0), 0) / Math.sqrt(gs.length);
    gScore = Math.min(1, gScore / 1.6);
    const rating = (rel.shikimori?.rating || rel.mal?.rating || 0) / 10;
    const tScore = p.types[rel.type?.description] || 0;
    const yScore = p.yearMean && rel.year ? Math.exp(-Math.abs(rel.year - p.yearMean) / 16) : 0.5;
    const total = 0.6 * gScore + 0.25 * rating + 0.1 * tScore + 0.05 * yScore;
    const common = gs.filter((x) => p.genres[x]).slice(0, 3);
    let best = null, bestN = 0;
    for (const [a, set] of likedSets) { const n = gs.filter((x) => set.has(x)).length; if (n > bestN) { best = a; bestN = n; } }
    // Причина — данными (жанры и похожий тайтл); текст по-своему языку собирает интерфейс
    out.push({ score: total, rel, reason: { genres: common, similar: best && bestN >= 2 ? best.title : null } });
  }
  return out.sort((a, b) => b.score - a.score);
}

/** Подборки «Потому что вам понравилось X»: [[понравившийся, [{score, rel}]]]. */
export function becauseOf(p, scored) {
  const rows = [];
  for (const liked of p.liked.slice(0, 3)) {
    const lg = new Set(liked.genres);
    if (lg.size < 2) continue;
    const picks = scored.filter((x) => (x.rel.genres || []).filter((g) => lg.has(g.name)).length >= Math.min(3, lg.size) - 1).slice(0, 12);
    if (picks.length >= 4) rows.push([liked, picks]);
  }
  return rows;
}

/** Главное о вкусе для текста разбора: {top: [жанры], year, type, typeShare}. */
export function summary(p) {
  if (p.empty) return null;
  const [type, share] = Object.entries(p.types)[0] || [];
  return { top: Object.keys(p.genres).slice(0, 3), year: p.yearMean ? Math.round(p.yearMean) : null, type: type || null,
    typeShare: share ? Math.round(share * 100) : 0 };
}

// ------------------------------------------------------------------ ИИ
export function extractJson(text) {
  text = String(text || "").trim();
  const cands = [text, ...(text.match(/\{[\s\S]*\}/g) || [])];
  for (const c of cands) {
    try {
      const d = JSON.parse(c);
      if (d?.choices) return extractJson(d.choices[0]?.message?.content);
      if (d && typeof d === "object") return d;
    } catch { /* не JSON — пробуем следующий кусок ответа */ }
  }
  return null;
}

const AI_LANGUAGE = { ru: "по-русски", en: "in English" };

/** lang — язык ответа ИИ (ru | en; для саха — по-русски: бесплатные модели якутский почти не знают). */
export function aiPrompt(p, pool, lang = "ru") {
  const liked = p.liked.slice(0, 15).map((a) => `${a.title} (${a.year || "?"}; ${(a.genres || []).slice(0, 4).join(", ")})`);
  const cands = pool.map((r, i) => `${i + 1}. ${title(r)} (${r.year || "?"}; ${(r.genres || []).map((x) => x.name).slice(0, 4).join(", ")})`);
  return `Ты эксперт по аниме. Сделай короткий разбор вкуса пользователя (4–6 предложений, ${AI_LANGUAGE[lang] || AI_LANGUAGE.ru}, дружелюбно) `
    + "и выбери из СПИСКА КАНДИДАТОВ 8 аниме, которые ему понравятся, с причиной до 12 слов.\n\n"
    + `Любимые жанры: ${Object.entries(p.genres).slice(0, 8).map(([k, v]) => `${k} (${Math.round(v * 100)}%)`).join(", ")}\n`
    + `Понравилось: ${liked.join("; ") || "мало данных"}\n\nСПИСОК КАНДИДАТОВ:\n${cands.join("\n")}\n\n`
    + 'Ответь ТОЛЬКО JSON: {"analysis": "...", "picks": [{"n": номер, "reason": "..."}]}';
}

/** {analysis, picks: [[rel, reason]]} или код ошибки ("format" | "unclear"), если ответ ИИ не годится. */
export function parseAi(text, pool, lang = "ru") {
  const d = extractJson(text);
  if (!d) return "format";
  const analysis = String(d.analysis || d["разбор"] || "");
  const picks = [];
  for (const x of d.picks || []) {
    let n = parseInt(x.n ?? x.id, 10) - 1;
    if (!(n >= 0 && n < pool.length)) n = pool.findIndex((r) => norm(title(r)) === norm(x.title));
    if (n >= 0 && n < pool.length && !picks.some(([r]) => r === pool[n])) picks.push([pool[n], x.reason || ""]);
  }
  // Иногда модель вместо ответа выдаёт рассуждения на другом языке — считаем это браком
  const letters = (analysis.match(lang === "en" ? /[a-z]/gi : /[а-яё]/gi) || []).length;
  if (!picks.length || letters < analysis.length * 0.3) return "unclear";
  return { analysis, picks };
}

File: anime_app/domain/taste.py
"""Анализ вкусов и рекомендации — только вычисления, без сети и базы.

Алгоритм (работает без ИИ, на ваших данных):
1. Сигналы: каждому тайтлу из библиотеки/истории — вес
   избранное +3, просмотрено +2, смотрю +1.5, хочу посмотреть +0.6, отложено +0.3, брошено −2,
   личная оценка (оценка − 5) / 2.5, досмотренные серии — до +1.5.
2. Профиль: взвешенные суммы по жанрам, типам и годам → нормированные предпочтения.
3. Оценка кандидата: 0.60 · совпадение по жанрам (косинус) + 0.25 · рейтинг
   + 0.10 · любимый тип + 0.05 · близость по годам. Объяснение — общие жанры и самый похожий
   тайтл из того, что вам понравилось.

ИИ получает только названия, жанры и годы — пишет разбор вкуса и выбирает из наших же кандидатов.
"""
from __future__ import annotations

import json
import math
import re
from collections import Counter, defaultdict
from typing import Callable

from ..core.i18n import service, t
from ..core.logging import get_logger
from .titles import norm, release_title

log = get_logger("taste")

STATUS_WEIGHT = {"completed": 2.0, "watching": 1.5, "planned": 0.6, "postponed": 0.3, "dropped": -2.0}
# Язык ответа ИИ — язык интерфейса. Саха модели пока не знают — разбор по-русски.
AI_LANGS = {"ru": "на русском языке", "sah": "на русском языке", "en": "на английском языке (in English)"}


def ai_lang(lang: str | None = None) -> str:
    return lang if (lang or "") in AI_LANGS else ("en" if service().locale == "en" else "ru")


def ai_system(lang: str | None = None) -> str:
    return (f"Ты помощник в приложении для просмотра аниме. Отвечай только валидным JSON {AI_LANGS[ai_lang(lang)]}, "
            "без рассуждений, пояснений и markdown.")


AI_SYSTEM = ai_system("ru")


def genres_of(release) -> list[str]:
    return [g.get("name") for g in (release or {}).get("genres") or [] if g.get("name")]


def type_of(release) -> str:
    return ((release or {}).get("type") or {}).get("description") or ""


def rating_of(release) -> float:
    for key in ("shikimori", "mal"):
        r = (release.get(key) or {}).get("rating")
        if r:
            return float(r)
    return 0.0


class TasteProfile:
    def __init__(self):
        self.titles: list[tuple[dict, float]] = []   # [(release, weight)]
        self.genres: dict[str, float] = {}           # жанр -> 0..1
        self.genres_raw: Counter = Counter()
        self.types: dict[str, float] = {}
        self.year_mean: float | None = None
        self.year_spread = 8.0
        self.status_counts: Counter = Counter()
        self.episodes = 0
        self.hours = 0.0
        self.liked: list[dict] = []                  # тайтлы с большим весом (для объяснений)
        self.disliked_genres: set[str] = set()

    @property
    def empty(self) -> bool:
        return not any(w > 0 for _, w in self.titles)

    def summary(self) -> str:
        top = ", ".join(f"{g} ({v:.0%})" for g, v in list(self.genres.items())[:8])
        types = ", ".join(f"{t} ({v:.0%})" for t, v in list(self.types.items())[:4])
        years = f"около {int(self.year_mean)} года" if self.year_mean else "неизвестно"
        return f"Жанры: {top}. Типы: {types}. Годы: {years}."

    def describe(self) -> str:
        """Разбор вкуса простыми словами."""
        if self.empty:
            return ""
        g = list(self.genres)
        parts = []
        if len(g) >= 2:
            parts.append(t("recs.taste_top3", a=g[0].lower(), b=g[1].lower(), c=g[2].lower()) if len(g) > 2
                         else t("recs.taste_top2", a=g[0].lower(), b=g[1].lower()))
        if self.year_mean:
            decade = int(self.year_mean) // 10 * 10
            parts.append(t("recs.taste_decade", decade=decade) if self.year_spread < 8
                         else t("recs.taste_years", year=int(self.year_mean)))
        if self.types:
            kind, v = next(iter(self.types.items()))
            parts.append(t("recs.taste_format", kind=kind.lower(), share=f"{v:.0%}"))
        if self.status_counts.get("dropped"):
            parts.append(t("recs.taste_dropped", n=self.status_counts["dropped"]))
        return " ".join(parts)


def build_profile(library: list[dict], watched: list[dict], releases: dict[int, dict]) -> TasteProfile:
    """library — строки списков (anime_id, status, favorite, score);
    watched — по тайтлам: (anime_id, eps, ms) — сколько серий досмотрено; releases — кэш карточек тайтлов."""
    p = TasteProfile()
    weights: defaultdict[int, float] = defaultdict(float)
    for entry in library:
        if entry.get("status"):
            p.status_counts[entry["status"]] += 1
        w = STATUS_WEIGHT.get(entry.get("status"), 0.0)
        if entry.get("favorite"):
            w += 3.0
        if entry.get("score"):
            w += (entry["score"] - 5) / 2.5
        weights[entry["anime_id"]] += w
    for r in watched:
        weights[r["anime_id"]] += min(r.get("eps") or 0, 12) / 8.0
        p.episodes += r.get("eps") or 0
        p.hours += (r.get("ms") or 0) / 3_600_000

    genre_w: Counter = Counter()
    type_w: Counter = Counter()
    year_sum = year_w = 0.0
    for aid, w in weights.items():
        rel = releases.get(aid)
        if not rel:
            continue
        p.titles.append((rel, w))
        gs = genres_of(rel)
        for g in gs:
            genre_w[g] += w / math.sqrt(len(gs))
            p.genres_raw[g] += 1
        if type_of(rel):
            type_w[type_of(rel)] += w
        if rel.get("year") and w > 0:
            year_sum += rel["year"] * w
            year_w += w
    pos = {g: v for g, v in genre_w.items() if v > 0}
    top = max(pos.values(), default=1.0)
    p.genres = dict(sorted(((g, v / top) for g, v in pos.items()), key=lambda x: -x[1]))
    p.disliked_genres = {g for g, v in genre_w.items() if v < -1}
    tpos = {t: v for t, v in type_w.items() if v > 0}
    ttot = sum(tpos.values()) or 1.0
    p.types = dict(sorted(((t, v / ttot) for t, v in tpos.items()), key=lambda x: -x[1]))
    if year_w:
        p.year_mean = year_sum / year_w
        dev = [abs((r.get("year") or p.year_mean) - p.year_mean) for r, w in p.titles if w > 0]
        p.year_spread = max(4.0, sum(dev) / len(dev)) if dev else 8.0
    p.liked = [r for r, w in sorted(p.titles, key=lambda x: -x[1]) if w >= 1.5][:12]
    return p


def candidate_queries(profile: TasteProfile, genre_ids: dict[str, int]) -> list[dict]:
    """Запросы каталога: любимые жанры по одному и парой + популярное и свежее."""
    top = [g for g in profile.genres if g in genre_ids][:4]
    queries = [{"genres": [genre_ids[g]], "sorting": "RATING_DESC", "limit": 40} for g in top]
    if len(top) >= 2:
        queries.append({"genres": [genre_ids[top[0]], genre_ids[top[1]]], "sorting": "RATING_DESC", "limit": 30})
    queries.append({"sorting": "RATING_DESC", "limit": 40})
    queries.append({"sorting": "FRESH_AT_DESC", "limit": 30})
    return queries


def score(profile: TasteProfile, releases: list[dict]) -> list[tuple[float, dict, str]]:
    """[(оценка, релиз, причина)] по убыванию."""
    liked_sets = [(r, set(genres_of(r))) for r in profile.liked]
    out = []
    for rel in releases:
        gs = genres_of(rel)
        if not gs:
            continue
        if profile.disliked_genres & set(gs) and len(profile.disliked_genres & set(gs)) >= len(gs) / 2:
            continue
        g_score = sum(profile.genres.get(g, 0.0) for g in gs) / math.sqrt(len(gs)) if profile.genres else 0
        g_score = min(1.0, g_score / 1.6)
        rating = rating_of(rel) / 10.0
        t_score = profile.types.get(type_of(rel), 0.0)
        y_score = 0.5
        if profile.year_mean and rel.get("year"):
            y_score = math.exp(-abs(rel["year"] - profile.year_mean) / (2 * profile.year_spread))
        total = 0.60 * g_score + 0.25 * rating + 0.10 * t_score + 0.05 * y_score
        common = [g for g in gs if g in profile.genres][:3]
        similar = max(liked_sets, key=lambda x: len(x[1] & set(gs)), default=(None, set()))
        reason = ""
        if common:
            reason = t("recs.reason_genres", genres=", ".join(common))
        if similar[0] and len(similar[1] & set(gs)) >= 2:
            reason += " · " + t("recs.reason_similar", title=release_title(similar[0]))
        out.append((total, rel, reason or t("recs.reason_popular")))
    out.sort(key=lambda x: -x[0])
    return out


def because_of(profile: TasteProfile, scored, limit: int = 12) -> list[tuple[dict, list[dict]]]:
    """Подборки «Потому что вам понравилось X»."""
    rows = []
    used: set = set()
    for liked in profile.liked[:3]:
        lg = set(genres_of(liked))
        if len(lg) < 2:
            continue
        picks = []
        for _s, rel, _ in scored:
            overlap = len(lg & set(genres_of(rel)))
            if overlap >= min(3, len(lg)) - 1 and rel["id"] not in used and rel["id"] != liked["id"]:
                picks.append(rel)
            if len(picks) >= limit:
                break
        if len(picks) >= 4:
            used.update(r["id"] for r in picks)
            rows.append((liked, picks))
    return rows


# ================================================================ ИИ
def extract_json(text):
    """JSON из ответа модели: голый, в ```-блоке или в обёртке OpenAI (choices → message → content)."""
    text = (text or "").strip()
    for candidate in (text, *re.findall(r"\{.*\}", text, re.S)):
        try:
            data = json.loads(candidate)
        except ValueError:
            continue
        if isinstance(data, dict) and data.get("choices"):
            content = ((data["choices"][0] or {}).get("message") or {}).get("content") or ""
            return extract_json(content)
        if isinstance(data, dict):
            return data
    return None


def ai_prompt(profile: TasteProfile, pool: list[dict], lang: str | None = None) -> str:
    liked = [f"{release_title(r)} ({r.get('year') or '?'}; {', '.join(genres_of(r)[:4])})"
             for r in profile.liked[:15]]
    cand = [f"{i + 1}. {release_title(r)} ({r.get('year') or '?'}; {', '.join(genres_of(r)[:4])})"
            for i, r in enumerate(pool)]
    return (
        "Ты эксперт по аниме. По данным пользователя сделай короткий разбор его вкуса (4–6 предложений, "
        f"{AI_LANGS[ai_lang(lang)]}, дружелюбно, без воды) и выбери из СПИСКА КАНДИДАТОВ 8 аниме, которые ему понравятся, "
        f"с короткой причиной для каждого (до 12 слов, {AI_LANGS[ai_lang(lang)]}).\n\n"
        f"Профиль: {profile.summary()}\n"
        f"Понравилось: {'; '.join(liked) or 'пока мало данных'}\n\n"
        "СПИСОК КАНДИДАТОВ:\n" + "\n".join(cand) + "\n\n"
        'Ответь ТОЛЬКО JSON без пояснений: {"analysis": "...", "picks": [{"n": номер, "reason": "..."}]}'
    )


def parse_ai_answer(text: str, pool: list[dict], title_of: Callable[[dict], str] = release_title,
                    lang: str | None = None):
    """(analysis, [(релиз, причина)]) или строка-ошибка, если ответ не годится (сам ответ — в журнал)."""
    data = extract_json(text)
    if data is None:
        log.info("ИИ: ответ не JSON: %s", (text or "").strip()[:300])
        return t("recs.ai_errors.format")
    analysis = data.get("analysis") or data.get("разбор") or data.get("анализ") or ""
    if not isinstance(analysis, str):
        analysis = json.dumps(analysis, ensure_ascii=False)
    raw_picks = data.get("picks") or data.get("recommendations") or next(
        (v for v in data.values() if isinstance(v, list)), [])
    picks: list[tuple[dict, str]] = []
    for p in raw_picks:
        if not isinstance(p, dict):
            continue
        n = -1
        for k in ("n", "номер", "id", "index"):
            try:
                n = int(p.get(k)) - 1
                break
            except (TypeError, ValueError):
                continue
        if not 0 <= n < len(pool):
            title = norm(p.get("title") or p.get("name") or p.get("название"))
            n = next((i for i, r in enumerate(pool) if title and norm(title_of(r)) == title), -1)
        if 0 <= n < len(pool) and pool[n] not in [x for x, _ in picks]:
            picks.append((pool[n], p.get("reason") or p.get("причина") or ""))
    # Иногда модель вместо ответа выдаёт свои рассуждения (по-английски) — для русского разбора это брак.
    cyr = sum(1 for ch in analysis if "а" <= ch.lower() <= "я")
    if not picks or (ai_lang(lang) != "en" and cyr < len(analysis) * 0.3):
        log.info("ИИ: непонятный ответ: %s", (text or "").strip()[:300])
        return t("recs.ai_errors.unclear")
    return analysis.strip(), picks
